modules/apps/29-fee/ibc_middleware_test.go

```diff
@@ -607,166 +607,220 @@ func (suite *FeeTestSuite) TestOnRecvPacket() {
+
+				// retrieve the relayer acc balance and add the expected ack fees
+				relayerAccBalance := sdk.NewCoins(suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), relayerAddr, sdk.DefaultBondDenom))
+				expPayeeAccBalance = relayerAccBalance.Add(packetFee.Fee.AckFee...)
 
-				expectedRelayerBalance = packetFee.Fee.AckFee
-				expectedBalance = expectedBalance.Add(packetFee.Fee.RecvFee...)
+				// retrieve the refund acc balance and add the expected recv fees and timeout fees
+				refundAccBalance := sdk.NewCoins(suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), refundAddr, sdk.DefaultBondDenom))
+				expRefundAccBalance = refundAccBalance.Add(packetFee.Fee.RecvFee...).Add(packetFee.Fee.TimeoutFee...)
 			},
-			false,
 			true,
+			func() {
+				// assert that the packet fees have been distributed
+				found := suite.chainA.GetSimApp().IBCFeeKeeper.HasFeesInEscrow(suite.chainA.GetContext(), packetID)
+				suite.Require().False(found)
+
+				relayerAccBalance := suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), relayerAddr, sdk.DefaultBondDenom)
+				suite.Require().Equal(expPayeeAccBalance, sdk.NewCoins(relayerAccBalance))
+
+				refundAccBalance := suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), refundAddr, sdk.DefaultBondDenom)
+				suite.Require().Equal(expRefundAccBalance, sdk.NewCoins(refundAccBalance))
+			},
 		},
 		{
-			"fail on no distribution by escrow account out of balance",
+			"fee distribution fails by escrow account out of balance",
 			func() {
-				packetID := channeltypes.NewPacketId(suite.path.EndpointA.ChannelConfig.PortID, suite.path.EndpointA.ChannelID, suite.chainA.SenderAccount.GetSequence())
 				err := suite.chainA.GetSimApp().BankKeeper.SendCoinsFromModuleToAccount(suite.chainA.GetContext(), types.ModuleName, suite.chainA.SenderAccount.GetAddress(), smallAmount)
 				suite.Require().NoError(err)
-
-				// expectedBalance and expectedRelayerBalance don't change
-				expectedBalance = originalBalance.Add(smallAmount...)
-				expectedRelayerBalance = sdk.NewCoins()
-
-				suite.chainA.GetSimApp().IBCFeeKeeper.SetFeesInEscrow(suite.chainA.GetContext(), packetID, types.NewPacketFees([]types.PacketFee{packetFee}))
-
-				expLocked = true
 			},
-			false,
 			true,
+			func() {
+				suite.Require().Equal(true, suite.chainA.GetSimApp().IBCFeeKeeper.IsLocked(suite.chainA.GetContext()))
+			},
+		},
+		{
+			"ack wrong format",
+			func() {
+				ack = []byte("unsupported acknowledgement format")
+			},
+			false,
+			func() {},
+		},
+		{
+			"invalid registered payee address",
+			func() {
+				payeeAddr := "invalid-address"
+				suite.chainA.GetSimApp().IBCFeeKeeper.SetPayeeAddress(
+					suite.chainA.GetContext(),
+					suite.chainA.SenderAccount.GetAddress().String(),
+					payeeAddr,
+					suite.path.EndpointA.ChannelID,
+				)
+			},
+			false,
+			func() {},
+		},
+		{
+			"application callback fails",
```

Comment:
should we still add a check that funds were distributed although the app callback failed?

Reply:
If the app callback fails it will bubble up and revert the whole tx. When the sdk routes the msg to the handler it caches the context.

[See here](https://github.com/cosmos/cosmos-sdk/blob/2b7aca73d2f8903f1ffa9f93c25793b7a3eb9d94/baseapp/baseapp.go#L693)